inst/include/common/information.hpp

```diff
@@ -20,81 +19,257 @@
+        typedef typename std::map<uint32_t, std::shared_ptr<fims::DataObject<T> > >::iterator data_iterator;  
+
+         //life history modules
+        std::map<uint32_t, std::shared_ptr<fims::RecruitmentBase<T> > > recruitment_models;//hash map to link each object to its shared location in memory
+        typedef typename std::map<uint32_t, std::shared_ptr<fims::RecruitmentBase<T> > >::iterator recruitment_model_iterator;
+  
+        std::map<uint32_t, std::shared_ptr<fims::SelectivityBase<T> > > selectivity_models;
+        typedef typename std::map<uint32_t, std::shared_ptr<fims::SelectivityBase<T> > >::iterator selectivity_models_iterator;
+  
+        std::map<uint32_t, std::shared_ptr<fims::GrowthBase<T> > > growth_models;
+        typedef typename std::map<uint32_t, std::shared_ptr<fims::GrowthBase<T> > >::iterator growth_models_iterator;
+  
+        //fleet modules
+        std::map<uint32_t, std::shared_ptr<fims::Fleet<T> > > fleets;
+        typedef typename std::map<uint32_t, std::shared_ptr<fims::Fleet<T> > >::iterator fleet_iterator;
+
+        //populations
+        std::map<uint32_t, std::shared_ptr<fims::Population<T> > > populations;
+        typedef typename std::map<uint32_t, std::shared_ptr<fims::Population<T> > >::iterator population_iterator;
+
+        //distributions
+        std::map<uint32_t, std::shared_ptr<fims::DistributionsBase<T> > >  distribution_models;
+        typedef typename std::map<uint32_t, std::shared_ptr<fims::DistributionsBase<T> > >::iterator distribution_models_iterator;
+
+        Information() {}
+
+        virtual ~Information() {}
+        
+        /**
+         * Returns a single Information object for type T.
+         *
+         * @return singleton for type T
+         */
+        static std::shared_ptr<Information<T> > GetInstance() {
+            if (Information<T>::fims_information == nullptr) {
+            Information<T>::fims_information =
+                std::make_shared<fims::Information<T> >();
+            }
+            return Information<T>::fims_information;
+        }
+
+        /**
+         * Register a parameter as estimable.
+         *
+         * @param p
+         */
+        void RegisterParameter(T& p) {
+            this->parameters.push_back(&p);
+        }
+            
+        /**
+         * Register a random effect as estimable.
+         *
+         * @param re
+         */
+        void RegisterRandomEffect(T& re) {
+            this->random_effects_parameters.push_back(&re);
+        }
+
+        /**
+         * Create the generalized stock assessment model that will evaluate the
+         * objective function. Does error checking to make sure the program has 
+         * all necessary components for the model and that they're in the right 
+         * dimensions.
+         *
+         * @return
+         */
+        bool CreateModel() {
+
+            bool valid_model = true;
```

Comment:
The logical setup for valid_model is well-coded and appears flawless.
I wonder if it may be useful to consider the ordering of the tests for
the valid_model flag? In particular, the current ordering in the code
is some thing like
catch
index for fleet
agecomp
selectivity
index likelihood
catch likelihood
agecomp likelihood
recruitment
growth
maturity
and I wonder if these 10 tests are listed in sequence that
matches a natural grouping of the valid_model tests?
Again, code looks fine as is but want to ask the question
about grouping of objects.

Along those same lines, I am wondering if there would 
be a unique code associated with the valid_model
failed tests to inform the developer or user where
the model did not satisfy the requirement? Or is this
something to be done in preprocessing in R?

Reply:
opened #254 to remind us to think about these questions in the future